src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,136 @@
+    private int size = 0;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+        } else if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index not valid");
+        } else {
+            if (size == elementData.length) {
+                grow();
+            }
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (elementData.length - size < list.size()) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            elementData[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
+        if (checkIndex(index)) {
+            return elementData[index];
+        }
         return null;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (checkIndex(index)) {
+            elementData[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = null;
+        if (checkIndex(index)) {
+            value = elementData[index];
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+            elementData[--size] = null;
+        }
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T value = null;
+        int index = -1;
+        if (null == element) {
+            for (int i = 0; i < size; i++) {
+                if (elementData[i] == null) {
+                    index = i;
+                    break;
+                }
+            }
+            value = checkRemoveIndex(index);
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (element.equals(elementData[i])) {
+                    index = i;
+                    break;
+                }
+            }
+            value = checkRemoveIndex(index);
+        }
+        return value;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index not valid");
+        }
+        return true;
+    }
+
+    private void grow() {
+        int newCapacity = size;
+        T[] newElementData;
+        if (size == elementData.length) {
```

Comment:
Seems we call grow() method only after size check were already performed in parent methods. Do we need additional check here?